contents/c++/rvalue.md

```diff
@@ -0,0 +1,169 @@
+* Speedup. If the function requires accessing of the value many times, having your own local copy of the object may be faster than having a pointer to the object and dereferencing it each time. This is especially true in C++ which encourages Value Semantics, with optimization techniques such as [copy elision](http://en.cppreference.com/w/cpp/language/copy_elision) in the compiler to help passing-by-value be faster.
+
+However, Value Semantics has [one major flaw](https://www.quora.com/What-are-the-drawbacks-of-pass-by-value-result):
+* Poor performance and scalability. If the function called is a read-only function, creating an additional copy will be an unnecessary consumption of memory and runtime. This is especially true when scaling up to pass around objects of larger size, as more memory and runtime is needed to create each copy.
+
+### Reference Semantics
+Reference (or pointer) semantics is another choice available for use in C++. C++ allows users to declare [pointers](http://www.cplusplus.com/doc/tutorial/pointers/) and [references](https://www.geeksforgeeks.org/references-in-c/) that point to the address of the object. As such, we can pass around these reference address to functions and declaration of other pointers, and all of them will refer to and use the same object and address.
+
+Some advantages of Reference Semantics include:
+* Improve performance and scalability. By passing around pointers and references to the same object, functions can just use the values in that object directly without having to create an extra copy, thus overcoming the major flaw of Value Semantics.
+* Allows interaction between objects and functions. Since the same object can be referred to by the different pointers, they can modify the same address values in different areas.
+
+However, Reference Semantics also comes with its own problems, which include:
+* Indeterminant behaviour in a multi-threaded environment. Since each thread will all be pointing to the same object, this will lead to unnecessary data races and indeterminant behaviour as these same objects are modified concurrently in each thread. This in turn requires extra work to synchronize the values between objects, which also slows down the program. More details of this can be found in the resources [Java Concurrency](java/JavaConcurrency.md) and [Java Synchronization](java/JavaSynchronization.md).
+* Unintended behaviour. If users are not careful, they can modify the values in the object using one pointer or reference, which will then be reflected by another pointer to the same object, which the user may not have intended. A good example of this problem can be found in this [article](http://www.drdobbs.com/cpp/optimization-calling-by-value-or-by-refe/232400151).
```

Comment:
Doesn't this point overlap with the `Indeterminant behaviour`?

Reply:
Hmm, does sound so...
Guess I will remove this section then :P

Reply:
Actually, `indeterminant` behaviour is different from `unintended`, since `indeterminant` means we can't determine what the result until we run the program, and it may produce different results each time.
But even if the result is `unintended`, in a sequential program the result should be deterministic.

Reply:
Ohh, then is it better to say `Unintended side-effects`?

Reply:
I'm thinking of writing this as `Harder to debug`, since if the behaviour is unintended, users will have a harder time trying to spot where the side-effected come from, especially if it is changed in distant place in the program?

Reply:
Sounds good to me :+1: